### PR TITLE
WebSocket: add extended logs

### PR DESCRIPTION
**Pull request by Amerousful**
https://github.com/gatling/gatling/issues/4306
***
Example
Code:
```scala
  val wsCheckCompleted = ws
    .checkTextMessage("Ws check [completed]")
    .matching(regex("COMPLETED"))

  val wsSendText = ws("Ws send text")
    .sendText("""{"auth":{"token":"blabla"}}""")
    .await(30 seconds)(wsCheckCompleted)
  ```
  
Logs:
  ```
  14:01:14.607 [DEBUG] i.g.h.a.w.WsLogger$ - Request 'Ws send text' failed for user 1: Check timeout
14:01:14.634 [DEBUG] i.g.h.a.w.WsLogger$ -
>>>>>>>>>>>>>>>>>>>>>>>>>>
Request:
Ws send text: KO Check timeout
=========================
Session:
Session(Example scenario,1,Map(gatling.http.ssl.sslContexts -> io.gatling.http.util.SslContexts@fa39da, gatling.http.cache.dns -> io.gatling.http.resolver.ShufflingNameResolver@74f3a467, gatling.http.cache.baseUrl -> https://httpbin.org, gatling.http.webSocket -> io.gatling.http.action.ws.fsm.WsFsm@5c577ea3),OK,List(),io.gatling.core.protocol.ProtocolComponentsRegistry$$Lambda$483/0x00000008005ba040@376ac420,io.netty.channel.nio.NioEventLoop@a68df9)
=========================
WebSocket check:
Ws check [completed]
=========================
WebSocket request:
{"auth":{"token":"blabla"}}
=========================
WebSocket received messages:
14:00:59.427 [0] -> {"status":"DISPATCHED"}
14:00:59.464 [1] -> {"status":"RECEIVED"}
14:00:59.502 [2] -> {"status":"RECEIVED"}
14:00:59.664 [3] -> {"status":"RECEIVED"}
14:01:00.209 [4] -> {"status":"RECEIVED"}
14:01:00.365 [5] -> {"status":"RECEIVED"}
14:01:00.444 [6] -> {"status":"FAILED"}
<<<<<<<<<<<<<<<<<<<<<<<<<

```

### CODE REVIEW

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsClosingState.scala [hunk 1]
```diff
@@ -33,11 +33,13 @@ final class WsClosingState(fsm: WsFsm, actionName: String, session: Session, nex
   }
 
   override def onTextFrameReceived(message: String, timestamp: Long): NextWsState = {
+    saveMessageToBuffer(message, timestamp)
     logUnmatchedServerMessage(session)
     NextWsState(this)
   }
 
   override def onBinaryFrameReceived(message: Array[Byte], timestamp: Long): NextWsState = {
+    saveMessageToBuffer(s"<<<BINARY CONTENT length=${message.length}>>>", timestamp)
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsClosingState.scala:42]: We should avoid this extra allocation when logging is disabled.

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsCrashedState.scala [hunk 2]
```diff
@@ -79,7 +80,7 @@ final class WsCrashedState(fsm: WsFsm, errorMessage: Option[String], val remaini
   ): NextWsState = {
     val loggedMessage = errorMessage match {
       case Some(mess) => s"Client issued a $frameType frame but WebSocket was already crashed: $mess"
-      case _          => "Client issued a $frameType frame but WebSocket was already closed"
+      case _          => s"Client issued a $frameType frame but WebSocket was already closed"
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsCrashedState.scala:83]: Nice catch, thanks!

File: src/docs/content/reference/current/http/websocket/index.md [hunk 3]
```diff
@@ -123,7 +123,7 @@ WebSocket support introduces new HttpProtocol parameters:
 You can inspect WebSocket traffic if you add the following logger to your logback configuration:
 
 ```xml
-<logger name="io.gatling.http.action.ws.fsm" level="DEBUG" />
+<logger name="io.gatling.http.action.ws.WsLogger" level="DEBUG" />
```
Comment (slandelle) [inline src/docs/content/reference/current/http/websocket/index.md:126]: We're losing the other logs then. Also, it's a change users might not notice when upgrading. We could move WsLogger in fsm instead.

**Review by slandelle — CHANGES_REQUESTED** [review]
Thanks a lot!
Could you please address my comments?

**Comment by Amerousful** [comment]
@slandelle Thank you for reviewing! I'm gonna fix

**Comment by slandelle** [comment]
@Amerousful Please re-request a review once you're done.
Screenshot 2022-08-10 at 22 52 23

**Comment by Amerousful** [comment]
@slandelle Got it.

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsIdleState.scala [hunk 4]
```diff
@@ -120,14 +129,15 @@ final class WsIdleState(fsm: WsFsm, session: Session, webSocket: WebSocket, prot
...
+    saveMessageToBuffer(s"<<<BINARY CONTENT length=${message.length}>>>", timestamp)
     // server push message, just log
     logUnmatchedServerMessage(session)
     NextWsState(this)
   }
 
   override def onWebSocketClosed(code: Int, reason: String, timestamp: Long): NextWsState = {
     // server issued close
-    logger.debug(s"WebSocket was forcefully closed ($code:$reason) by the server while in Idle state")
+    logger.debug(s"WebSocket was forcefully closed ($code/$reason) by the server while in Idle state")
```
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsIdleState.scala:140]: I decided to unified message like in `WsPerformingCheckState` and `WsClosingState`:

Screenshot 2022-08-18 at 18 38 09

File: gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+  private def messageBufferToRaw(messageBuffer: ArrayBuffer[(Long, String)]): String = {
+    messageBuffer.zipWithIndex
+      .map {
+        // time [index] -> message
+        case (mes, ind) =>
+          val time = new SimpleDateFormat("HH:mm:ss.SSS").format(mes._1)
+          s"$time [$ind] -> ${mes._2}"
+      }
+      .toSeq
```
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala:44]: Not anymore. After changed `messageBuffer` type...

File: gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import io.gatling.core.session.Session
+import io.gatling.http.engine.response.HttpTracing
+import io.gatling.http.util._
+import io.gatling.netty.util.StringBuilderPool
+
+import com.typesafe.scalalogging.StrictLogging
+
+object WsLogger extends StrictLogging {
+
+  private val loggingStringBuilderPool = new StringBuilderPool
```
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala:34]: @slandelle Hey! Can you give me some advice?
Do I need to create an object of this WsLogger in `WsConnectRequestBuilder` and pass it into `WsConnect`?

**Comment by Amerousful** [comment]
@slandelle I don't sure completely about function naming `wsLogsPiling`, so any suggestions?

File: gatling-http/src/main/scala/io/gatling/http/util/package.scala [hunk 7]
```diff
@@ -107,5 +110,18 @@ package object util extends LazyLogging {
...
       }
+
+    def appendWebsocketInboundMessages(messageBuffer: Seq[(Long, String)]): jl.StringBuilder = {
+      buff.append(
+        messageBuffer.zipWithIndex
+          .map { case ((timestamp, message), index) =>
+            val time = bufferFormat.format(timestamp)
+            s"$time [$index] -> $message"
+          }
+          .mkString("\n")
```
Comment (phiSgr) [inline gatling-http/src/main/scala/io/gatling/http/util/package.scala:121]: If you use a loop to repeatedly `append` to the `StringBuilder`, you can save the allocation of

- the `List` with indices,
- the `List` of `String`s
- the final concatenated `String`.

Also it looks like you need the final new line anyway.

Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/util/package.scala:121]: @phiSgr What do I need to use instead of `map`?

Comment (phiSgr) [reply gatling-http/src/main/scala/io/gatling/http/util/package.scala:121]: Use [`foreach`](https://www.scala-lang.org/api/2.13.x/scala/collection/Seq.html#foreach[U](f:A=%3EU):Unit) for effects.

This principle also applies in many other languages like [JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) and [Java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-).

Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/util/package.scala:121]: now I see) thx)
Oh, I know about `foreach`, but I don't know why I forgot it 🙈

**Review by slandelle — CHANGES_REQUESTED** [review]
@Amerousful I'm very sorry, I forgot to hit the "Submit" button...

File: gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java [hunk 8]
```diff
@@ -1139,6 +1139,15 @@ public HttpProtocolBuilder wsAutoReplySocketIo4() {
     return new HttpProtocolBuilder(wrapped.wsAutoReplySocketIo4());
   }
 
+  /**
+   * Store received message to buffer and print to stdout
+   *
+   * @return a new HttpProtocolBuilder instance
+   */
+  public HttpProtocolBuilder wsLogsPiling() {
```
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java:1147]: > @Amerousful I'm very sorry, I forgot to hit the "Submit" button...
That's all right!

[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java:1147]: > I really don't think this feature should be controlled at the protocol level. It's only meant for debugging purpose, so just like other logging features, it should be controlled by enabled slf4 loggers, eg `io.gatling.http.engine.response.HttpTracing`.

Now I see, got it. I'll fix it

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsIdleState.scala [hunk 9]
```diff
@@ -82,6 +86,9 @@ final class WsIdleState(fsm: WsFsm, session: Session, webSocket: WebSocket, prot
     // actually send message!
     val now = clock.nowMillis
     webSocket.sendFrame(new BinaryWebSocketFrame(Unpooled.wrappedBuffer(message)))
+    val requestMessage =
+      if (HttpTracing.IS_HTTP_TRACE_ENABLED || HttpTracing.IS_HTTP_DEBUG_ENABLED) Some(s"<<<BINARY CONTENT length=${message.length}>>>") else None
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsIdleState.scala:90]: redundant: if trace is enabled, debug is => `if (HttpTracing.IS_HTTP_DEBUG_ENABLED)` is enough.

File: src/docs/content/reference/current/cheat-sheet.md [hunk 10]
```diff
@@ -1505,6 +1505,9 @@ content:
 
           - keyword: wsAutoReplySocketIo4
             description: Enable partial support for Engine.IO v4 - Gatling will automatically respond to server ping messages (<i>2</i>) with pong (<i>3</i>).
+            
+          - keyword: wsLogsPiling
```
Comment (slandelle) [inline src/docs/content/reference/current/cheat-sheet.md:1509]: This change should be reverted.

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/StatsProcessorWS.scala [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+import io.gatling.commons.util.StringHelper.Eol
+import io.gatling.core.session.Session
+import io.gatling.core.stats.StatsEngine
+import io.gatling.http.response.{ HttpResult, Response }
+import io.gatling.http.util._
+import io.gatling.jdk.util.StringBuilderPool
+
+import com.typesafe.scalalogging.StrictLogging
+
+sealed abstract class StatsProcessor extends StrictLogging {
```
[the remark this answers is not in the document]
Comment (slandelle) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/StatsProcessorWS.scala:29]: Where do you use this? I don't see those classes instantiated anywhere. This is dead code replaced by WsLogger, right?

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsFsm.scala [hunk 12]
```diff
@@ -43,16 +44,19 @@ class WsFsm(
     private[fsm] val httpEngine: HttpEngine,
     private[fsm] val httpProtocol: HttpProtocol,
     eventLoop: EventLoop,
-    private[fsm] val clock: Clock
+    private[fsm] val clock: Clock,
+    private[fsm] val wsLog: WsLogger
```
Comment (slandelle) [inline gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/WsFsm.scala:48]: wsLog => wsLogger

**Review by slandelle — CHANGES_REQUESTED** [review]

File: gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (slandelle) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/WsLogger.scala]: Sorry, missed this comment. Let go first with your design and I might play with it and iterate. Have to poc design ideas in PR reviews.

File: gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/StatsProcessorWS.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/StatsProcessorWS.scala:29]: I have no clue what is it... 🙈

[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http/src/main/scala/io/gatling/http/action/ws/fsm/StatsProcessorWS.scala:29]: I have no clue what is it... 🙈 Just deleted it

File: src/docs/content/reference/current/cheat-sheet.md [hunk 13]
```diff
@@ -1505,7 +1505,7 @@ content:
 
           - keyword: wsAutoReplySocketIo4
             description: Enable partial support for Engine.IO v4 - Gatling will automatically respond to server ping messages (<i>2</i>) with pong (<i>3</i>).
-
+
```
Comment (slandelle) [inline src/docs/content/reference/current/cheat-sheet.md:1508]: extra spaces :)

**Review by slandelle — CHANGES_REQUESTED** [review]
Almost there 😂

File: src/docs/content/reference/current/cheat-sheet.md (hunk 13 above)
Comment (Amerousful) [reply src/docs/content/reference/current/cheat-sheet.md:1508]: oh, shame on me) 1 sec

**Review by slandelle — APPROVED** [review]

**Comment by slandelle** [comment]
cherry-picked, thanks a lot!

**Comment by Amerousful** [comment]
Thank you too!

File: gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java]: @slandelle hey again) 

We forgot about this! My point is that now we can't separate `http` and `ws` logs. Precisely speaking, with ```
<logger name="io.gatling.http.action.ws" level="TRACE"/>``` we will see `ws` logs, but store message to buffer won't work

[the remark this answers is not in the document]
Comment (Amerousful) [reply gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java]: And seems like we need to create `WsTracing` inside `io.gatling.http.action.ws`

[the remark this answers is not in the document]
Comment (slandelle) [reply gatling-http-java/src/main/java/io/gatling/javaapi/http/HttpProtocolBuilder.java]: > My point is that now we can't separate http and ws logs.

I don't think we want to.

**Pull request closed without merging**